packages/hash/datastore/postgres/migration/1648547815337_base.ts

```diff
@@ -0,0 +1,613 @@
+      },
+    },
+    {
+      ifNotExists: true,
+    },
+  );
+
+  pgm.addConstraint("entity_types", "entity_type_unique_constraint", {
+    unique: ["account_id", "name"],
+  });
+
+  pgm.createTable(
+    "entity_type_versions",
+    {
+      entity_type_version_id: {
+        type: "uuid",
+        notNull: true,
+        primaryKey: true,
+      },
+      account_id: {
+        type: "uuid",
+        notNull: true,
+        references: {
+          name: "accounts",
+        },
+      },
+      entity_type_id: {
+        type: "uuid",
+        notNull: true,
+        references: {
+          name: "entity_types",
+        },
+      },
+      properties: {
+        type: "jsonb",
+        notNull: true,
+      },
+      updated_by_account_id: {
+        type: "uuid",
+        notNull: true,
+        references: {
+          name: "accounts",
+        },
+      },
+      updated_at: {
+        type: "timestamp with time zone",
+        notNull: true,
+        comment: stripNewLines(`
+          Versioned types are never mutated, so the updated_at time always matches the
+          created_at time. Non-versioned types may be mutatated in-place, and the
+          updated_at column changes when a mutation is made.
+        `),
+      },
+    },
+    {
+      ifNotExists: true,
+    },
+  );
+
+  pgm.createIndex("entity_type_versions", "entity_type_id", {
+    name: "entity_type_versions_entity_type_id",
+    ifNotExists: true,
+  });
+
+  /**
+   * @todo: does this play well with citus?
+   * JSONB index for componentIds, this will speed up checking for an entity with a specific componentId property.
+   */
+  pgm.createIndex(
+    "entity_type_versions",
+    "((properties ->> 'componentId'::text))",
+    {
+      name: "entity_type_version_component_id",
+      ifNotExists: true,
+    },
+  );
+
+  /**
+   * @todo: Does this play well with citus?
+   * JSONB GIN index on allOf field of JSON schema. jsonb_path_ops option on the GIN index optimizes for @> operations.
+   * https://www.postgresql.org/docs/9.5/gin-builtin-opclasses.html
+   * Used for accessing children of (types that inherit from) an EntityType.
+   */
+  pgm.createIndex("entity_type_versions", "(properties -> 'allOf')", {
+    method: "gin",
+    name: "entity_type_versions_prop_allOf",
+    ifNotExists: true,
+  });
+
+  /**
+   * @todo: Does this play well with citus?
+   * The $id field of the JSON schema is used to find schemas referred to in other schemas by $ref
+   * For example, this is used to find all parent EntityTypes (types this EntityType inherits).
+   * The contents of allOf is traversed and matched on the property field '$id'
+   */
+  pgm.createIndex(
+    "entity_type_versions",
+    "((properties ->> '$id'::text))", // @todo: ((properties -> 'allOf') jsonb_path_ops)
+    {
+      name: "entity_type_version_prop_id",
+      ifNotExists: true,
+    },
+  );
+
+  pgm.createTable(
+    "entities",
+    {
+      account_id: {
+        type: "uuid",
```

Comment:
One annoyance is that this isn't (partially) typed by the library. That would make DX slightly nicer, but oh well!

Reply:
Yeah had the same thought ... we could try re-declaring some of the modules types, but not sure this is worth the effort. If there's a typo, the migration will throw an error so we'd catch it in CI.